### PR TITLE
CloudWatch GetMetricData support

### LINKED ISSUE #11487: [Feat] CloudWatch GetMetricData support
https://docs.aws.amazon.com/AmazonCloudWatch/latest/APIReference/API_GetMetricData.html
We can use Math expression!!

### PR DESCRIPTION
**Pull request by mtanda**
Fixes https://github.com/grafana/grafana/issues/11487

I introduce two fields, "Id" and "Expression".

These fields are pass as MetricDataQuery, and other parameters are filled by existing fields.
https://docs.aws.amazon.com/AmazonCloudWatch/latest/APIReference/API_MetricDataQuery.html

If user uses templating expansion in dimension value, the Id is populated by templating value.
https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/using-metric-math.html

The populated Id is useful for calculating sum by `SUM(METRICS("<Id>"))`.

### CODE REVIEW

**Comment by mtanda** [comment]
https://github.com/aws/aws-sdk-go/issues/1892 is released and applied to this PR.
Ready for merge.

**Comment by mtanda** [comment]
@bergquist Would you check this changes?

File: public/app/plugins/datasource/cloudwatch/partials/query.parameter.html [hunk 1]
```diff
@@ -31,6 +31,13 @@
 	</div>
 </div>
 
+<div class="gf-form-inline">
+	<div class="gf-form">
+		<label class="gf-form-label query-keyword width-8">Expresion</label>
```
Comment (daniellee) [inline public/app/plugins/datasource/cloudwatch/partials/query.parameter.html]: Spelling - should be `Expression`

**Review by daniellee — COMMENTED** [review]
Hi @mtanda how do you use this feature? Could you give some examples of how to use math in the expression.

We will also need to add some documentation for this feature (I can help!)

**Comment by mtanda** [comment]
I provide two examples.

Example 1: calculating DynamoDB capacity usage rate
image

image

Example 2: calculating Average of CPU Utilization of All EC2 instance
image

File: public/app/plugins/datasource/cloudwatch/partials/query.parameter.html [hunk 2]
```diff
@@ -31,18 +31,31 @@
...
 
-<div class="gf-form-inline">
+<div class="gf-form-inline" ng-if="target.statistics.length === 1">
 	<div class="gf-form">
-		<label class="gf-form-label query-keyword width-8">
+		<label class=" gf-form-label query-keyword width-8 ">Id</label>
+		<input type="text " class="gf-form-input " ng-model="target.id " spellcheck='false' ng-model-onblur ng-change="onChange() ">
+	</div>
+	<div class="gf-form max-width-30 ">
+		<label class="gf-form-label query-keyword width-7 ">Expresion</label>
```
Comment (gcacace) [inline public/app/plugins/datasource/cloudwatch/partials/query.parameter.html]: Typo here: `expresion`

Comment (mtanda) [reply public/app/plugins/datasource/cloudwatch/partials/query.parameter.html]: I forgot to fix it. Thanks for pointing out.

**Comment by daniellee** [comment]
@mtanda I tested now and had some problems:

- If there are template variables in the query then the query id gets a unique uid added to the end: https://github.com/mtanda/grafana/blob/fb0cd5a29d84165d261f800cbcd9df7a21478c1b/public/app/plugins/datasource/cloudwatch/datasource.ts#L397

    So this query does not work as the query id for the first query gets set to `network_packets_inaS1iMGU4YTQ0Nw00`:
    image

    If I change the expression to `network_packets_inaS1iMGU4YTQ0Nw00/2` then it works or if I take out all the template variables from the query then it works.

- If the second query does not have an id then it does not work and the error message does not help much: 
    ```
    InvalidParameter: 2 validation error(s) found.
    - minimum field size of 1, GetMetricStatisticsInput.MetricName.
    - minimum field size of 1, GetMetricStatisticsInput.Namespace.
    ```

**Comment by mtanda** [comment]
@daniellee Thank you for feedback!

> So this query does not work as the query id for the first query gets set to network_packets_inaS1iMGU4YTQ0Nw00:

For this issue, I add check the variable is `multi` or not.
If variable is not `multi`, `network_packets_in / 2` should be valid expression, so don't generate unique id, user can use `network_packets_in` in expression.
If variable is `multi`, datasource should generate unique id, and user need to use expression like `AVG(METRICS("network_packets_in")) / 2`. This is expected behavior, because user set variable as `multi`.

> If the second query does not have an id then it does not work and the error message does not help much:

Improve error message. It is little tricky, so will need documentation...

**Comment by mdgbayly** [comment]
@mtanda Thanks for working on this - this is something we would really like to see supported by Grafana. We've had to switch back to Cloudwatch for some dashboards due to inability to use math expressions with `ConsumedReadCapacityUnits`. 

Regarding using this feature for DynamoDB capacity utilizations, do you know if it will be possible in Grafana to utilize the current period in the calculation.

E.g. above you show calculating used_per_sec as `used / 60` (i realize you are setting min period to 60).
In Cloudwatch math expressions you can include the period in the expression.
`used / PERIOD(used)`. 

Does the Grafana support allow using any Cloudwatch Math functions in an expression?

**Comment by mtanda** [comment]
@mdgbayly 

> Does the Grafana support allow using any Cloudwatch Math functions in an expression?

Yes, you can use all Math functions.

To calculate used_per_sec, your example is correct!
`used / PERIOD(used).`

**Comment by daniellee** [comment]
@mtanda I’m on vacation this week (and last week). Will merge this on Monday after testing again.

**Comment by mtanda** [comment]
@daniellee No problem. I know it. Thanks for informing me :-)

**Comment by daniellee** [comment]
Merging now. There is one more slightly strange thing that some people might get stuck on - that id's have to start with a lowercase letter. From the docs:

`You can change the value of Id. It can include numbers, letters, and underscore, and must start with a lowercase letter`

Not sure what to do about it though. Maybe should an info popover for the id field. What do you think Mtanda?

**Comment by mtanda** [comment]
@daniellee Thanks for merging!
I didn't notice id's limitation. I'll add popover or input validator.

**Comment by kadern0** [comment]
Hi @mtanda,
great new feature! I'm trying to use it but unfortunately if I enter any value in the Id field all datapoints disappear. 
Any ideas?

**Comment by mtanda** [comment]
Hi @kadern0,

At first, you or Grafana running instance should have privilege to call new `cloudwatch:GetMetricData` API.
https://docs.aws.amazon.com/AmazonCloudWatch/latest/APIReference/API_GetMetricData.html

Please check your IAM policy.
If IAM policy is correct and Grafana report error, please share error message.

**Comment by rickbowden** [comment]
@mtanda 
Thanks so much for working on this. I've spent quite some time attempting to make aggregated panels using CloudWatch datasource, this will make it so much better.
Thanks again.

**Comment by gipeshka** [comment]
Hi @mtanda,
Thanks for the feature!

However I face the same problem @kadern0 had, as soon as I put any value in `Id` field data disappears. Query response from grafana also has no data:
```
{"results":{}}
```
I've checked the privilege, it's there. Calls from aws cli using grafana user retrieve the data.

**Comment by mtanda** [comment]
Hi @gipeshka,

Would you capture your query editor and share with me?
And if you have a time, would you try for simple EC2 metrics (e.g. `CPUUtilization`)?

**Comment by gipeshka** [comment]
Hi @mtanda.

Thank you for a quick reply. I was able to solve the issue. With the new api apparently one should pay more attention to the naming (first I used "-" symbol in the id, grafana didn't complain, however I got no data) and to general query structure. Maybe some additional validation with more explicit errors would help the case.

**Comment by gipeshka** [comment]
@mtanda another quick question, does grafana support alerts on the calculated metrics/metrics with `Id`?
When I try to test such an alert grafana always returns `no_data`.

**Comment by mtanda** [comment]
@gipeshka Now I understand Grafana alerting model...
Currently, alerting doesn't work.

When alerting expression evaluated, only one expression is passed to datasource.
exp A (id=a): `CPUUtlization`
exp B (id=b): AVG(a)

And then, set alert rule for exp B, only `AVG(a)` is passed to datasource.

**Comment by mtanda** [comment]
Graphite datasource can handle such nested query alert.

I'll try to fix metric math alerting.
https://github.com/grafana/grafana/issues/9917

**Comment by tphambgs** [comment]
Hi @mtanda, 

I'm trying to use repeat math expression for a multi value dimension:

image

In the screenshot above, $tables is a list of DynamoDB tables. The query generates a unique `used` ID for each table but doesn't repeat this ID in the expression. 

Is there a use-case where we wouldn't want to repeat this ID?

**Comment by mayukuse24** [comment]
Hi @mtanda 

Will the support for multiple statistics be added anytime soon or is it in progress?

**Comment by mayukuse24** [comment]
> At first, this feature has limitation, it doesn't support multiple statistics in one target.
> User should specify only one statistics (e.g. Average).
> I improve query editor in last commit, if user specify multiple statistics, extra field is disappeared.
> 
> Unless user set extra field, `Id` and `Expression`, the result is same as current one.
> 
> `Id` and `Expression` is explained in following docs.
> https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/using-metric-math.html
> 
> User need to set `Id` for each time series, and then use it in `Expression`.
> 
> By using this feature, user can calculate like `used / total = used rate`.
> To use with templating feature, generate unique id for each expanded target.
> https://github.com/grafana/grafana/pull/11624/files#diff-5be6ea1dd05a4db39f000439bbca4f6aR397
> User can simply using aggregate function like "AVG" with expanded target.
> 
> If user don't want to show the intermediate result, user can hide it by clicking `eye` icon in query editor, but it still available in expression.
> 
> I provide two examples.
> 
> Example 1: calculating DynamoDB capacity usage rate
> calculate `ConsumedReadCapacityUnits per second usage`.
> image
> 
> divide `ConsumedReadCapacityUnits per second usage` by `ProvisionedReadCapacityUnits`, get used rate.
> image
> 
> Example 2: calculating Average of CPU Utilization of All EC2 instance
> image

Hi @mtanda

Will the support for multiple statistics be added anytime soon or is it in progress?

**Comment by adithyamenon** [comment]
> @daniellee Thank you for feedback!
> 
> > So this query does not work as the query id for the first query gets set to network_packets_inaS1iMGU4YTQ0Nw00:
> 
> For this issue, I add check the variable is `multi` or not.
> If variable is not `multi`, `network_packets_in / 2` should be valid expression, so don't generate unique id, user can use `network_packets_in` in expression.
> If variable is `multi`, datasource should generate unique id, and user need to use expression like `AVG(METRICS("network_packets_in")) / 2`. This is expected behavior, because user set variable as `multi`.
> 
> > If the second query does not have an id then it does not work and the error message does not help much:
> 
> Improve error message. It is little tricky, so will need documentation...

I don't think this serves the purpose entirely.

For example, in my case, I am trying to plot the 'ConsumedWriteCapacityUnits' from DynamoDB keeping the list of tables as variable - '$tables'. In this case, it wouldn't make sense if I get the average of 'ConsumedWriteCapacityUnits' of the selected multiple DynamoDB tables. I would essentially want individual values for the multiple tables selected without the generated unique id as mentioned by @daniellee 

Am I right here or am I missing something? @mtanda 

I am currently using Grafana Version 5.4.2 (commit: d812109, branch: HEAD).

Screenshot 2019-04-02 at 1 12 31 PM

**Comment by mtanda** [comment]
I misunderstand about some part of GetMetricData API.
This will fix issue.
https://github.com/grafana/grafana/pull/16383

**Pull request merged**